README.md
xbmc-rai-on-demand
==================

Xbmc plugin that permits to watch all rai on demand tv shows on the xbmc media center. 
All contents are the same available on the rai website www.rai.tv in the on demand section. 

Content rights remain reserved and owned by Rai.

Installation instructions [EN]
=========================
Easy way:
- Step 1: Download source code from github
- Step 2: Create a zip file of the folder "plugin.video.rai-on-demand" (Note: on mac os x does not use the built-in archive utility, it creates an unreadable zip file)
- Step 3: Go into XBMC and install add-on from zip file from the settings menù

Manual way:
- Step 1: Download source code from github
- Step 2: Copy the folder "plugin.video.rai-on-demand" into the addons folder of xbmc (Note: each OS has its specific path)
- Step 3: Restart xbmc and you will be able to see the addon under the video section

Guida installazione [IT]
=========================

Installazione semplice:
- Step 1: Scarica i sorgenti da github
- Step 2: Crea un file zip della cartella "plugin.video.rai-on-demand" (Nota: su mac os x non utilizzare l'utility di compressione di sistema, questa crea zip file non installabili da xbmc)
- Step 3: Vai nel menù impostazioni di xbmc e seleziona installa l'addon da zip file

Installazione manuale:
- Step 1: Scarica i sorgenti da github
- Step 2: Copiate la cartella "plugin.video.rai-on-demand" dentro la cartella addons di xbmc (Nota: ogni OS ha il suo path)
- Step 3: Riavvia xbmc and troverai l'addon nella sezione video



plugin.video.rai-on-demand/changelog.txt
Version 1.0.0
  First version of the plugin
Version 1.0.1
  Solved annoying problem of reproduction failure due to xbmc timeout

plugin.video.rai-on-demand/test.py
'''
Created on Nov 21, 2012

@author: flavio
'''
import request,sys

request.indexReq()
print request.indexShowList('/dl/RaiTV/programmi/page/Page-86e553b7-5d8c-4375-809c-1bf1c879c1c3.html', 0)

def getParams():
    print "ARGUMENTS: "
    print sys.argv
    
    resultDict = {}
    
    if len(sys.argv) > 1 and len(sys.argv[3]) > 1:
        paramList = sys.argv[3][1:].split('&')
        for elem in paramList:
            token = elem.split('=')
            if len(token) == 2:
                resultDict[token[0]] = token[1]
    
    return resultDict

print getParams()

plugin.video.rai-on-demand/main.py
'''
Created on Nov 23, 2012

@author: flavio
'''

import urllib,sys,request,xbmcWrapper,json
import urllib2

pluginId = 0

#addon id - name of addon directory
_id='plugin.video.rai-on-demand'
#resources directory
_resdir = "special://home/addons/" + _id + "/resources" #add our library to python search path
sys.path.append( _resdir + "/lib/")

def loggerMethod(top,log):
    print '#############################'
    print top + '-' + log

def addLetterIndex():
    ''' This method creates the list of initial letter of all shows '''
    global pluginId,base_url
    
    letterList = request.letterIndexReq()
    # Add a list of letter
    for letter in letterList:
        xbmcWrapper.addFolder(pluginId,1,letter,{'letter': letter})
    xbmcWrapper.endOfContent(pluginId)
    
def addTvShowsByLetter(paramDict):
    ''' This methods create the list of shows starting with letter '''        
    showList = request.showsWithLetterReq(paramDict['letter'])   
    # Add the list of shows starting with the letter chosen
    for elem in showList:
        xbmcWrapper.addFolder(pluginId,2,elem.get('title'),{'title': elem.get('title'), 'iconImage': elem.get('image'), 'linkDemand': elem.get('linkDemand')})
    xbmcWrapper.endOfContent(pluginId)    
        
def addTvShowsCategories(paramDict):
    ''' This methods create the list of shows video categories ''' 
    # LinkDemand is an URL that passed as a parameter form the command line and had been escaped so need to be recovered       
    url = urllib.unquote_plus(paramDict['linkDemand'])         
    categoryList = request.showVideoCategories(url)  
    #Add the list of categories for the chosen show
    for elem in categoryList:
        xbmcWrapper.addFolder(pluginId,3,elem[1],{'title': paramDict['title'], 'categoryName': elem[1], 'contentSet-Id': elem[0], 'page': str(0)})
    xbmcWrapper.endOfContent(pluginId)  

def addTvShowsCategoryEpisodes(paramDict):
    ''' This methods create the list of video available for the previously chosen show and category '''  
    episodeList = request.showsEpisodeList(paramDict['contentSet-Id'], int(paramDict['page']))   
    # Add the list of video for the specified category
    for elem in episodeList['list']:
        video = elem.get('h264') 
        
        if video == None or len(video) < 1:
            video = elem.get('wmv')
        if video == None or len(video) < 1:
            video = elem.get('mediaUri')
        if video:
            #req = urllib2.Request(video)
            #req.add_header('User-Agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11')
            #response = urllib2.urlopen(req)
            #video = response.geturl()
            print '*********' + video
            xbmcWrapper.addVideoItem(pluginId, elem.get('name'), video, 'http://www.rai.tv/'+elem.get('image'), 'http://www.rai.tv/'+str(elem.get('image_medium')))
    
    # Hanlde the next page indicator
    if int(paramDict['page'])+1 < int(episodeList['pages']):
        xbmcWrapper.addFolder(pluginId,4,'Prossima Pagina',{'title': paramDict['title'], 'contentSet-Id': paramDict['contentSet-Id'], 'page': str(int(paramDict['page'])+1)})
    
    xbmcWrapper.endOfContent(pluginId)    

def getParams():
    ''' Utility method that returns all params passed to xbmc as a dictionary'''
    print "ARGUMENTS: "
    print sys.argv
    
    resultDict = {}
    
    if len(sys.argv) > 1 and len(sys.argv[2]) > 1:
        paramList = sys.argv[2][1:].split('&')
        for elem in paramList:
            token = elem.split('=')
            if len(token) == 2:
                resultDict[token[0]] = token[1]
    
    return resultDict
        

paramsDict=getParams()
mode=None
try:
        mode=int(paramsDict['mode'])
except:
        pass
    
loggerMethod("Mode", str(mode))


# Mode = None - First start - List of letters
if mode == None: 
    print "Create groups"
    addLetterIndex()

# Mode = 1 - Letter chosen - List of TV shows starting with defined letter
elif mode == 1: 
    print "Create shows staring with " + paramsDict['letter']
    addTvShowsByLetter(paramsDict)

# Mode = 2 - Tv show chosen - List of video categories available    
elif mode == 2: 
    print "Create video categories for: " + paramsDict['title']
    addTvShowsCategories(paramsDict)
    
# Mode = 2 - Tv show chosen - List of episode available for that show      
elif mode == 3: 
    print "Create episode index for: " + paramsDict['title'] + ' page: ' + str(paramsDict['page'])
    addTvShowsCategoryEpisodes(paramsDict)

# Mode = 3 - Tv show chosen - List of episode available for that show for next pages    
elif mode == 4: 
    print "Create episode index for: " + paramsDict['title'] + ' page: ' + str(paramsDict['page'])
    addTvShowsCategoryEpisodes(paramsDict)


plugin.video.rai-on-demand/cache.py
'''
Created on Nov 21, 2012

@author: flavio
'''
import xbmc,os,sys,time

tempPath = xbmc.translatePath('special://temp/')


def getFileCache(name,freshnessSec):
    ''' Get a file from cache if exists and checks that is enough fresh '''
    result = None
    try:
        f = open(tempPath + name)
        # get the timestamp to understand if the file is still valid
        try:
            t = f.readline().rstrip()
            print 'localtime: ' +str(time.time())
            print 'savedTime: ' + t
            print 'difference: ' + str(time.time()-float(t))
            if (time.time()-float(t)) < freshnessSec:
                result = f.read()
                f.close()
                print 'getShowsWithLetterCache - RESULT CACHED'
        except:
            os.remove(tempPath + 'showIndex.txt')
    except:
        pass
    return result

def saveFileCache(name,file):
    try:
        os.remove(tempPath + "XBMC_RAI_" + name)
    except:
        pass
    f = open(tempPath + "XBMC_RAI_" + name,'w')
    f.write(str(time.time())+'\n')
    f.write(file)
    f.close()

def clearFileCache():
    for f in os.listdir(tempPath):
        if "XBMC_RAI_" in f:
            os.remove(tempPath+f) 
        

plugin.video.rai-on-demand/request.py
'''
Created on Nov 20, 2012
This module handle request to URLs and use file for caching response 

@author: flavio
'''

import http,re,cache,json
from config import base_url

def letterIndexReq():
    ''' Returns the list of initial letter for shows available'''
    resultList = ['0/9']
    for letter in range(ord('A'), ord('Z')+1):
        resultList.append(chr(letter))
    
    return resultList

def showsWithLetterReq(letter):
    ''' Returns the list of shows starting with the letter passed in JSON format'''
    index = cache.getFileCache('showsIndex.txt',56700)  # 56700 means that the cache file must be younger than 12 hours
    if index == None:
        index = http.getPage('http://www.rai.tv/dl/RaiTV/programmi/ricerca/ContentSet-6445de64-d321-476c-a890-ae4ed32c729e-darivedere.html')
        cache.saveFileCache('showsIndex.txt',index)

    indexList = json.loads(index)
    
    filteredList = []
    for elem in indexList:
        if elem['title'].startswith(letter) or (letter[0].isdigit() and elem['title'][0].isdigit()):
            filteredList.append(elem)
    
    return filteredList

def showVideoCategories(showUrl):
    ''' Returns the list of video categories as tuple, with their contentSet, available for the show chosen'''
    global base_url
    
    showPage = http.getPage(base_url + showUrl)
    
    match=re.compile('<li class="Video ">\n(.+?)\n</li>').findall(showPage)
    
    result = []
    for line in match:
        category = re.compile('<a.+?id="(.+?)".+?>(.+?)</a>').findall(line)
        if category:
            result += category

    return result

def showsEpisodeList(contentId,page):
    ''' Returns the list of episodes of the chosen show'''
    
    videoPage = cache.getFileCache(contentId + '-' + str(page) + '.txt', 56700)
    
    if videoPage == None:
        videoPage = http.getPage('http://www.rai.tv/dl/RaiTV/programmi/json/liste/' + contentId + '-json-V-' + str(page) + '.html')
        cache.saveFileCache(contentId + '-' + str(page) + '.txt',videoPage)

    videoJSON = json.loads(videoPage)
    
    return videoJSON
    
    

plugin.video.rai-on-demand/xbmcWrapper.py
'''
Created on Nov 22, 2012

@author: flavio
'''
import sys,xbmcplugin,xbmcgui,urllib

def createModeUrl(mode, paramDict):
    urlParams = 'mode=' + str(mode)
    
    for dictItem in paramDict.items():
        param = ''
        try:
            param = urllib.quote_plus(dictItem[1].encode('ascii', 'ignore'))
        except:
            param = dictItem[1].encode('ascii', 'ignore')
        urlParams += '&' + dictItem[0] + '=' + param
        
    # url = "plugin://" + self._pluginName + "?url=" + item[1]
    #u=sys.argv[0]+"?url="+urllib.quote_plus(url)+"&mode="+str(mode)+"&name="+urllib.quote_plus(name)
    url = sys.argv[0] + '?' + urlParams

    print "xxxDEBUGxxx createModeUrl ---> Mode:" + str(mode) + " url: " + url

    return url

def addFolder(pluginId, mode, name, paramDict):
    url = createModeUrl(mode,paramDict)

    image="DefaultFolder.png"
    if 'image' in paramDict:
        image = paramDict['image']
        
    item=xbmcgui.ListItem(label=name, iconImage=image)
    # Add the item to xbmc
    xbmcplugin.addDirectoryItem(pluginId, url, item, isFolder = True)

def endOfContent(pluginId):
    xbmcplugin.endOfDirectory(pluginId)

def addVideoItem(pluginId, name, videoUrl, iconImage, thumbnailImage):
    item=xbmcgui.ListItem(name, iconImage=str(iconImage), thumbnailImage=str(thumbnailImage))
    item.setInfo(type="Video", infoLabels={ "Title": name })
    xbmcplugin.addDirectoryItem(pluginId, videoUrl, item)

def addVideoItemWithMode(pluginId, mode, name, videoUrl, iconImage, thumbnailImage):
    url = createModeUrl(mode, {'title':name, 'videoUrl':str(videoUrl), 'iconImage':str(iconImage), 'thumbnailImage':str(thumbnailImage)})
    item = xbmcgui.ListItem(name, iconImage=str(iconImage), thumbnailImage=str(thumbnailImage))
    item.setInfo( type="Video", infoLabels={ "Title": name } )
    xbmcplugin.addDirectoryItem(pluginId, url, item, isFolder = True)


plugin.video.rai-on-demand/default.py
'''
Created on Nov 23, 2012

@author: flavio
'''

import urllib,urllib2,sys,request,xbmcWrapper,json,cache,xbmc,xbmcgui

pluginId = int(sys.argv[1])

#addon id - name of addon directory
_id='plugin.video.rai-on-demand'
#resources directory
#_resdir = "special://home/addons/" + _id + "/resources" #add our library to python search path
#sys.path.append( _resdir + "/lib/")

def loggerMethod(top,log):
    '''use for debug'''
    print '############################## LOGGER ##############################'
    print top + '-' + log

def addLetterIndex():
    ''' This method creates the list of initial letter of all shows '''
    global pluginId,base_url
    
    letterList = request.letterIndexReq()

    # Add a list of letter
    for letter in letterList:
        xbmcWrapper.addFolder(pluginId,1,letter,{'letter': letter})
    xbmcWrapper.endOfContent(pluginId)
    
def addTvShowsByLetter(paramDict):
    ''' This methods create the list of shows starting with letter '''        
    showList = request.showsWithLetterReq(paramDict['letter'])   
    # Add the list of shows starting with the letter chosen
    for elem in showList:
        xbmcWrapper.addFolder(pluginId,2,elem.get('title'),{'title': elem.get('title'), 'image': elem.get('image'), 'linkDemand': elem.get('linkDemand')})
    xbmcWrapper.endOfContent(pluginId)    
        
def addTvShowsCategories(paramDict):
    ''' This methods create the list of shows video categories ''' 
    # LinkDemand is an URL that contains category list of the specific tv show      
    url = paramDict['linkDemand']         
    categoryList = request.showVideoCategories(url)  
    #Add the list of categories for the chosen show
    for elem in categoryList:
        xbmcWrapper.addFolder(pluginId,3,elem[1],{'title': paramDict['title'], 'categoryName': elem[1], 'contentSet-Id': elem[0], 'page': str(0)})
    xbmcWrapper.endOfContent(pluginId)  

def addTvShowsCategoryEpisodes(paramDict):
    ''' This methods create the list of video available for the previously chosen show and category '''  
    episodeList = request.showsEpisodeList(paramDict['contentSet-Id'], int(paramDict['page']))   
    # Add the list of video for the specified category
    for elem in episodeList['list']:
        video = elem.get('h264') 
        
        if video == None or len(video) < 1:
            video = elem.get('wmv')
        if video == None or len(video) < 1:
            video = elem.get('mediaUri')
        if video:
            xbmcWrapper.addVideoItemWithMode(pluginId, 5, elem.get('name'), video, 'http://www.rai.tv/'+elem.get('image'), 'http://www.rai.tv/'+str(elem.get('image_medium')))
    
    # Hanlde the next page indicator
    if int(paramDict['page'])+1 < int(episodeList['pages']):
        xbmcWrapper.addFolder(pluginId,4,'Prossima Pagina',{'title': paramDict['title'], 'contentSet-Id': paramDict['contentSet-Id'], 'page': str(int(paramDict['page'])+1)})
    
    xbmcWrapper.endOfContent(pluginId)

def getVideoUrlandPlay(paramDict):
    ''' This methods get the video url following redirect and set it to play '''
    req = urllib2.Request(urllib.unquote_plus(paramDict['videoUrl']))
    req.add_header('User-Agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11')
    response = urllib2.urlopen(req)
    video = response.geturl()
    
    print "VIDEO is: " + video

    videoInfo=xbmcgui.ListItem(urllib.unquote_plus(paramDict['title']), paramDict['iconImage'], paramDict['thumbnailImage'])
    videoInfo.setInfo( type="Video", infoLabels={ "Title": urllib.unquote_plus(paramDict['title']) } )
    # Play the video
    xbmc.Player().play(video, videoInfo)    

def getParams():
    ''' Utility method that returns all params passed to xbmc as a dictionary'''
    resultDict = {}
    
    if len(sys.argv) > 1 and len(sys.argv[2]) > 1:
        paramList = sys.argv[2][1:].split('&')
        for elem in paramList:
            token = elem.split('=')
            if len(token) == 2:
                resultDict[token[0]] = urllib.unquote_plus(token[1])
    
    return resultDict
        
# New plugin call
print '########################## NEW PLUGIN CALL ################################'
print "ARGUMENTS: "
print sys.argv
paramsDict=getParams()
print "PARAMS PARSED"
print paramsDict
print '############################################################################'

mode=None
try:
    mode=int(paramsDict['mode'])
except (TypeError, KeyError):
    pass
except:
    print "********* GENERIC ERROR EXCEPTION TO BE VERIFIED *********"
    print sys.exc_info()
    cache.clearFileCache()

print 'MODE - ' + str(mode)

# Mode = None - First start - List of letters
if mode == None: 
    print "Create groups"
    addLetterIndex()

# Mode = 1 - Letter chosen - List of TV shows starting with defined letter
elif mode == 1: 
    print "Create shows staring with " + paramsDict['letter']
    addTvShowsByLetter(paramsDict)

# Mode = 2 - Tv show chosen - List of video categories available    
elif mode == 2: 
    print "Create video categories for: " + paramsDict['title']
    addTvShowsCategories(paramsDict)
    
# Mode = 3 - Tv show chosen - List of episode available for that show      
elif mode == 3: 
    print "Create episode index for: " + paramsDict['title'] + ' page: ' + str(paramsDict['page'])
    addTvShowsCategoryEpisodes(paramsDict)

# Mode = 4 - Tv show chosen - List of episode available for that show for next pages    
elif mode == 4: 
    print "Create episode index for: " + paramsDict['title'] + ' page: ' + str(paramsDict['page'])
    addTvShowsCategoryEpisodes(paramsDict)

# Mode = 5 - Video chosen - Follow redirects and get the last video url   
elif mode == 5: 
    print "Get video from url: " + paramsDict['videoUrl'] + ' and play'
    getVideoUrlandPlay(paramsDict)
